### PR TITLE
ci: GitHub Actions to validate a Maven BOM

### LINKED ISSUE #5922: How might we enhance validation on BOMs
How might we enhance validation on BOMs?

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #5922 

It worked in https://github.com/googleapis/gapic-generator-java/pull/1629

Todo: 
- Merge this pull request 5928.
- Update the branch name in https://github.com/googleapis/gapic-generator-java/pull/1629 to "main"
- Add another check in https://github.com/googleapis/google-cloud-java with "main"

### CODE REVIEW

**Comment by suztomo** [comment]
The check detected the invalid entries in the previous gapic-libraries-bom successfully https://github.com/googleapis/java-cloud-bom/actions/runs/4703742327/jobs/8342595777?pr=5928:

`[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.253 s
[INFO] Finished at: 2023-04-14T20:41:59Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal on project bom-validation-canary-project: Could not resolve dependencies for project com.google.cloud:bom-validation-canary-project:jar:0.0.1-SNAPSHOT: The following artifacts could not be resolved: com.google.analytics.api.grpc:grpc-google-analytics-admin-v1alpha:jar:0.24.0, com.google.analytics.api.grpc:grpc-google-analytics-admin-v1beta:jar:0.24.0, com.google.analytics.api.grpc:proto-google-analytics-admin-v1alpha:jar:0.24.0, com.google.analytics.api.grpc:proto-google-analytics-admin-v1beta:jar:0.24.0, com.google.analytics.api.grpc:grpc-google-analytics-data-v1beta:jar:0.25.0, com.google.analytics.api.grpc:grpc-google-analytics-data-v1alpha:jar:0.25.0, com.google.analytics.api.grpc:proto-google-analytics-data-v1beta:jar:0.25.0, com.google.analytics.api.grpc:proto-google-analytics-data-v1alpha:jar:0.25.0, com.google.area120.api.grpc:grpc-google-area120-tables-v1alpha1:jar:0.18.0, com.google.area120.api.grpc:proto-google-area120-tables-v1alpha1:jar:0.18.0: Could not find artifact com.google.analytics.api.grpc:grpc-google-analytics-admin-v1alpha:jar:0.24.0 in central (https://repo.maven.apache.org/maven2) -> [Help 1]`

File: .github/workflows/full-convergence-check.yaml [hunk 1]
```diff
@@ -12,16 +12,17 @@ jobs:
...
+        java-version: 11
+        cache: maven
     - run: java -version
     - name: Install BOMs
       run: |
         mvn -B -V -ntp install
     - name: Ensure the members of the Libraries BOM exist in Maven Central
-      run: |
-        mvn -B -V -ntp verify -Dtest="BomContentTest#testLibrariesBomReachable"
-      working-directory: tests
```
Comment (suztomo) [inline .github/workflows/full-convergence-check.yaml:24]: This new check can replace testLibrariesBomReachable

File: .github/workflows/full-convergence-check.yaml [hunk 2]
```diff
@@ -12,16 +12,17 @@ jobs:
...
+      uses: ./tests/validate-bom
+      with:
+        bom-path: libraries-bom/pom.xml
```
Comment (suztomo) [inline .github/workflows/full-convergence-check.yaml:25]: This is how to use the action within the repository.

Until https://github.com/googleapis/java-cloud-bom/pull/5927 is merged, this check fails.

File: .github/workflows/ci-release-note-generation.yaml [hunk 3]
```diff
@@ -12,6 +12,7 @@ jobs:
       with:
         distribution: zulu
```
Comment (burkedavison) [inline .github/workflows/ci-release-note-generation.yaml:13]: Reminder about Temurin

File: tests/validate-bom/action.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      if [ ! -r "${{ inputs.bom-path }}" ]; then
+        echo "The input bom-path ${{ inputs.bom-path }} is not readable"
+        exit 1
+      fi
+      
+      bom_absolute_path=$(realpath "${{ inputs.bom-path }}")
+      # Before this "cd", the working directory is the repository that calls
+      # this action. To use validate-bom classes, it needs to change directory
+      # to the directory that defines this action.
+      cd ${{ github.action_path }}
```
Comment (burkedavison) [inline tests/validate-bom/action.yml:34]: Is this change in directory going to persist for actions that occur after this one? Consider whether a `cd -` might needed.

File: tests/validate-bom/action.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      echo "Compiling CreateBomCanaryProject.java in $(pwd)"
+      mvn -V -ntp compile
+      echo "Running CreateBomCanaryProject with ${bom_absolute_path}"
+      mvn -V -ntp -B exec:java -DoutputPath=/tmp/bom-validation -DbomPath="${bom_absolute_path}"
```
Comment (burkedavison) [inline tests/validate-bom/action.yml:38]: If this action is used twice in a repo with two different BOMs, will the first invocation pollute the project setup or results of the second? (I think the answer is: this is fine. However, if GitHub were to run two BOM validations concurrently, there might be a conflict.) If we're unsure, consider making the BOM's output path calculated from the BOM-under-tests' name.

File: tests/validate-bom/src/main/java/com/google/cloud/CreateBomCanaryProject.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    String outputPathProperty = System.getProperty("outputPath");
+    checkNotNull(outputPathProperty, "System property outputPath should not be null");
+    Path outputProjectDirectory = Paths.get(outputPathProperty);
+    String bomPathProperty = System.getProperty("bomPath");
+    checkNotNull(bomPathProperty, "System property bomPath should not be null");
+    Path bomPath = Paths.get(bomPathProperty);
+
+    Bom bom;
+    try {
+      bom = Bom.readBom(bomPath);
```
Comment (burkedavison) [inline tests/validate-bom/src/main/java/com/google/cloud/CreateBomCanaryProject.java:37]: Neat

**Review by burkedavison — APPROVED** [review]

File: .github/workflows/ci-release-note-generation.yaml (hunk 3 above)
Comment (suztomo) [reply .github/workflows/ci-release-note-generation.yaml]: Done

File: tests/validate-bom/action.yml (hunk 5 above)
Comment (suztomo) [reply tests/validate-bom/action.yml:40]: GitHub Actions do not run in parallel on one filesystem.

**Pull request merged**